core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -477,7 +477,7 @@ private[spark] class BlockManager(
             }
           } else {
             // Otherwise, we also have to store something in the memory store
-            if (!level.deserialized || !asBlockResult) {
+            if (!level.deserialized && !asBlockResult) {
```

Comment:
I neglected to do this earlier because this entire block of code is being completely rewritten in another one of my PRs.